rust/tw_ss58_address/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+        Self::from_u16(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct SS58Address {
+    key: Vec<u8>,
+    network: NetworkId,
+}
+
+impl SS58Address {
+    const CHECKSUM_SIZE: usize = 2;
+    const KEY_SIZE: usize = 32;
+    const SS58_PREFIX: &'static [u8] = b"SS58PRE";
+
+    fn extract_network(bytes: &[u8]) -> AddressResult<(usize, u16)> {
+        if bytes.len() == 0 {
+            return Err(AddressError::MissingPrefix);
+        }
+
+        match bytes[0] {
+            0..=63 => Ok((1, bytes[0] as u16)),
+            64..=127 if bytes.len() >= 2 => {
+                let lower = (bytes[0] << 2) | (bytes[1] >> 6);
+                let upper = bytes[1] & 0b0011_1111;
+                Ok((2, (lower as u16) | ((upper as u16) << 8)))
+            },
+            _ => Err(AddressError::UnexpectedAddressPrefix),
+        }
+    }
+
+    fn encode_network(network_id: NetworkId) -> Vec<u8> {
+        let network = network_id.value();
+        match network {
+            0..=63 => {
+                vec![network as u8]
+            },
+            _ => {
+                let first = ((network & 0b0000_0000_1111_1100) as u8) >> 2;
+                let second =
+                    ((network >> 8) as u8) | ((network & 0b0000_0000_0000_0011) as u8) << 6;
+                vec![first | 0b01000000, second]
+            },
+        }
+    }
+
+    fn compute_expected_checksum(decoded: &[u8]) -> Vec<u8> {
+        // XXX: tried using tw_hash::blake2::blake2_b, but it did not produce the correct result
+        let mut ctx = Blake2b512::new();
+        ctx.update(Self::SS58_PREFIX);
+        ctx.update(decoded);
+
+        let mut bytes = ctx.finalize().to_vec();
+        bytes.truncate(Self::CHECKSUM_SIZE);
+        bytes
+    }
+
+    pub fn from_str(repr: &str) -> AddressResult<Self> {
+        let decoded = base58::decode(repr, base58::Alphabet::BITCOIN)
+            .map_err(|_| AddressError::FromBase58Error)?;
+
+        let (network_prefix_len, network) = Self::extract_network(&decoded)?;
+
+        if network_prefix_len + Self::KEY_SIZE + Self::CHECKSUM_SIZE != decoded.len() {
+            return Err(AddressError::FromBase58Error);
+        }
+
+        let expected_checksum =
+            Self::compute_expected_checksum(&decoded[..decoded.len() - Self::CHECKSUM_SIZE]);
+        let checksum = &decoded[decoded.len() - Self::CHECKSUM_SIZE..];
+
+        if expected_checksum != checksum {
+            return Err(AddressError::InvalidChecksum);
+        }
+
+        Ok(Self {
+            key: decoded[network_prefix_len..network_prefix_len + Self::KEY_SIZE].to_owned(),
+            network: NetworkId::new_unchecked(network),
+        })
+    }
+
+    pub fn from_public_key(key: &PublicKey, network: u16) -> AddressResult<Self> {
+        let network = NetworkId::try_from(network)?;
+
+        let bytes = match key {
+            PublicKey::Ed25519(k) => k.as_slice(),
+            _ => return Err(AddressError::InvalidInput),
```

Comment:
`AddressError::PublicKeyTypeMismatch` suits here too.
Also, personally I prefer to take a strict public key type in the address implementation, so `tw_keypair::ed25519::sha512::PublicKey`, and do the `tw::PublicKey -> tw_keypair::ed25519::sha512::PublicKey` conversion in `entry.rs` implementation.
For example,
https://github.com/trustwallet/wallet-core/blob/de287c8edc1bd5c5c8aec1b1185d38c9fd1f9b5d/rust/chains/tw_greenfield/src/entry.rs#L66-L69